### DISCUSSION TITLE
Upscaling Flare Art

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [flareteam/flare-game#895] [Q&A] [answered]
https://www.youtube.com/watch?v=Fix6u4pksrg
^This, for Flare?

The Flare Engine now has the ability to host at least 4 times the graphical resolution of the Legacy Flare Engine and this can lead the way forward towards Flare HD. 64-bit executable for the Flare Engine is now available across all major PC Platforms. Effectively giving the Flare Engine with the 64-bit executable 18.4 Exabytes of usable RAM over the previous 4.2 Gigabyte of usable RAM from the legacy 32-bit executable.

The 64-bit executable for Flare empowers the Flare Engine to be able to operate with effectively 4K+ support. Meaning that the original Flare art assets can now be enlarged x 4 in scale to achieve 4 times the pixel density with stable Flare Engine Performance where this was simply impossible with the Legacy Flare Engine and the hard limit of usable RAM of 4.2 Gigabytes. Of which the 32-bit executable was insufficient to host enlarged x 4 Flare art assets without crashing half way through rendering due to insufficient memory errors.

The 64-bit executable for all major PC systems can bring the Flare Engine from a technological era of the 2000's with pixelated legacy graphics running at low frames per second and a 32-bit executable; into a technological era of the 2020's with 4K clear graphics with high frames per second and a 64-bit executable. The 64-bit executable Flare Engine is proven to be able to run with 4 times the original resolution of the legacy 32-bit executable Flare Engine. With the 64-bit executable Flare Engine is proven with high frame rate that each individual legacy pixel can now become 4 pixels in the same screen real-estate / pixel density; thus proving the viability of high resolution Flare Engine operation. I call this 64-bit executable enlarged x 4 art asset version of the Flare Engine 'Flare x 4'.
![AAAAAAAAAAAAAAAA!!! Flare x 4 64-bit Success!](https://user-images.githubusercontent.com/86871191/151306141-b66d0427-1103-4241-a8d7-bc99ac58cbbd.png)
The above is an initial prototype of a successful test running the Windows 64-bit executable of the Flare Engine successfully rendering at 60 frames per second with an enlarged x 4 Flare player character avatar sprite. With also an enlarged x 4 Fantasy core art asset tileset for an area map of 1000 x 1000 tiles in scale (for 1,000,000 tiles in total map scale area). With each tile being 256x128 pixels in dimension; 4 times scale of the legacy original Flare art scale at a scale of 64 x 32 pixels.

The Flare Engine performance to host Flare x 4 (and the path toward Flare HD) is there and proven to work. Yet the graphical detail is not yet properly harnessed. As such the prototype above can have 4 times the detail and could effectively host high resolution 4K graphical detail art assets if they existed for Flare. Of which I currently understand that there are no Flare x 4 and or Flare 4K and or Flare HD art assets available for the creative community to harness this 64-bit executable high resolution graphical power.

Should there be an art campaign undertaken to Upscale Flare Art to improve the art presentation quality of the original art? This Upscaling Flare Art process could be simpler than making new renders of original Flare Art blender files. This also could serve as an interim / in-between step for improving the Flare graphics and could potentially lead to Flare HD Version 1.0. Would there be community interest in such an art campaign to Upscale Flare Art? Also is there any experience / advice / interest with Upscaling Art within the creative community?

**Comment by dorkster** [comment]
This has been an open topic for a long time now: https://github.com/flareteam/flare-game/issues/209

My thoughts:

- We still want to support people playing on 32-bit CPUs. So we can't just blow things up to the point where a 64-bit CPU would be *required*. That being said, packing and splitting up spritesheets would probably keep us within 32-bit limitations anyway. So a 64-bit executable is not the game changer you think it is.
- We want to support less performant hardware in general. I was doing some playing on a Raspberry Pi 4 model B (4GB) recently, and was able to play at 1600x900@60FPS. However, I needed to decrease the Maximum Render Size setting in Flare to 480p to keep it at a smooth 60.
- AI upscaling is great and all, but I'd rather only use it when needed. We certainly have some assets where we don't have access to a larger source. But for 3D models, we should be able to render those out at any resolution. Which brings me to my next point...
- The biggest hurdle to properly doing this is fixing the workflow for the 3D files. I think I mentioned this before, but artists should be able to press a single button to run a script that will output the spritesheet/tileset and any accompanying text files in a form ready to use in Flare.
- I would rather stick with Clint's plan to do 2x instead of 4x scale. The reason is a combination of performance and the quality of the original source files. Past a certain point, I think there are diminishing returns.

**Reply by WithinAmnesia** [reply]
My fear is with future proofing.

If we work towards only what we stand on currently or what is within convenient reach; it will be outdated and of limited future use. I have high aims to make an industry competitive video game in the sense that it can live up to the common expectations of most people today and into the 2030's (hopefully beyond that as well). I cannot grow this vision if I am not actively aiming into the distance proactively.

Baldur's Gate Enhanced Edition suffered greatly from the retention of pixelated graphics even in the 2010's. The developer staff wanted to make it an HD version yet they lacked the source files and had to have other areas of focus to compensate. I feel as a sense of dread within my heart that this approach of staying where we are, even after over a decade of hard work could be suicide; for growth for a video game wishing to fulfill the expectations of common people in the 2020's and 2030's. As in to be of a complacent heart and mind that has settled with less than what we could become and use admittedly dated and obsolete presentation of the video game. I understand that there are games with ~'bad graphics' / low resolution art and do fine (minecraft, terraria, stardew valley etc). Although these Flare related video games by their very inception and nature of gameplay will most always being living under the shadow of Diablo II and Path of Exile.

I know that there is the potential to compete with the high quality experience offered by these two franchise titans of the industry in with Diablo and Path of Exile. The creative community can do so without hundreds of developer staff and millions of dollars to realize their vision; effectively democratizing video game development. Which is pretty massive in and of it self. Yet I don't think we are far off from being at a point where within reason the Flare Engine / Fantasy core / the creative community can have the opportunity to go toe to toe with the titans of the industry and win.

To slay the proverbial demons plaguing the video game industry. To earn the liberty to be a creative community free from creative slavery of corporate greed, abuse and mismanagement of community trust, and the overt, shameless, honourless, indignant and open betrayal of gaming community's faith and good will for the creation of video games worth playing. As what is currently the case of our status quo across the world yet does not have to be as such. So long as there is a collective hunger of a creative community to become more than the sum of their collective parts and through their struggle and hopefully ultimately shared victory over the suffering of the bad old days, win a new dawn of new creative hope to inspire and empower the entire world; for today and tomorrow, rich or poor.

I dream and hope to share a dream of generations forever free and creatively empowered in a world with the collective heart of hearts closer to becoming something greater than one's self and perhaps close to what we were always meant to be: In a world full of omnipresent darkness, together as one rise above our inner doubts, inhibitions and terror of the unknown to stand defiantly tall and able as an unyielding shining beacon that of the collective fire of humanity; that burns in the heart of every person, mighty or meek that has ever existed and will come to be in our shared human experience.

We are so close to be at a point where the tools can exist to make great games with low input and high output to creatively empower the creative community. So they don't need to be slaves to abusive corporations and their hopes and dreams can flourish and thrive to be made by themselves with a proper chance at success to realize their passions, visions and dreams. On their own terms and with upholding their own personal sovereignty and creative freedom. For the tools to build these great games could be forever free and open to the entire world; rich or poor.

You would not need to even live in the richest parts of the world to have a shot to make the game you dream of as well. Hell if someone builds a masterpiece here even in the heart of darkness of the Congo a person could be inspired and start making great video games and have the creative power to do so. Like how this guy with nothing but endless adversity, a beautiful dream and an unyielding passion went and earned success in making his own movies https://www.youtube.com/watch?v=sy0OOVTmsJI&ab_channel=VICE . Admittedly they are not perfect, but it is quite the accomplishment from having worse than nothing to make something that can inspire the whole world. 

I understand that Flare cannot throw the baby out with the bathwater nor do I advocate for gutting 32-bit support entirely. I am focused on these ~4K / Flare HD graphics for they are at a point where it captures the beauty and the spirit of these video games so amazingly and anything beyond this amount of detail for 2.5D games in like you Justin @dorkster said results in diminishing returns. 4K seems to be the point where anything beyond that is almost impossible for a person to notice beyond having their eyes right next to the screen like in VR. I kind of personally internalize the struggle for Flare HD as the last ultimate battle to win to realize the mighty empowering force that the Flare creative tool set has the very potential to manifest into being. A mighty creatively empowering force that can thrive to become greater than the sum of its collective parts for the creative community to make great video games that could stand the test of time and last decades as timeless classics.

I do not expect the battle for Flare HD to be easy, short, convenient nor accomplished within one fell swoop. It is mostly likely a community campaign with multiple small yet critically important steps combined together to create big strides towards an ultimately shared victory for the creative community. Yet if such a victory for Flare HD is indeed won, it can be a victory that lasts for decades and does away with the pixelated limitations of our previous generation of technology. It can stand as the ultimate example perhaps of the dream version of what was originally desired and worked towards when the first 2.5D graphics were pioneered; yet lacked the computing power to realize that vision and fulfill such an artistic beauty.

This potential inspires myself to try harder and reach that critical mass where the beauty of the graphics that can be found in ~4K and with the efficient work flow of tiles and human readability of the Flare Engine meet; a sort of creative near-paradise. I want the Flare creative tool set to be in a place where the graphics will age well over time, the game play is deep and robust and with the good will and faith of the creative community. A forever free creative near-paradise that is home to timeless classics birthed from the passions and dreams of the entire world; rich or poor together as one.

In my heart of hearts I feel that Flare HD is worth doing. I realize that upon such a path towards Flare HD I will be deeply challenged and some challenges that I face I cannot reasonably overcome readily by myself without great daring and power of will. Yet somehow I know that I can find a way even if it takes time. Compound interest is a mighty force to be reckoned with. So bit by bit, piece by piece until it all clicks and that beauty and creative power can be realized for the creative community.

[It is hard to describe a feeling given by a dream at times; yet it is readily apparent to understand and feel deeply on a personal level.](https://www.youtube.com/watch?v=ufQUxoidxkM&ab_channel=Boston-Topic) I work towards my dreams of realizing a medieval fantasy world that haunts myself continually. Yet in my near-never ending battles throughout the decades to realize this ever-haunting dream to be manifested for the heart of hearts of the entire world. Within this deep struggle of mine fighting with myself and with my adversities to manifest my creative power. That I know is latent deep within the depths of my soul and of my heart of hearts. In my deeply scarring personal pains and of my worldly shaping character forging challenges along my own journey. I have found that there are so many people who dream the same beautiful visions that stir the very depths of my heart of hearts. Yet unfortunately so they do not manifest their dreams for they lack the reasonable opportunity to utilize a creative tool set to empower them to do so; even as one solitary soul. So I feel that as I battle near-ever onwards towards my own goals to manifest into reality and to share with the entire world my dreams that near-ever insistently haunt myself so of a series of video games worth playing and experiencing at least once in a lifetime. I wish it so that any success that I encounter I hope to be able to share with the creative community; and hopefully make the world a better and more reasonable and hopeful place then when I found it originally.

Video games are like waking dreams in many ways. With the graphics being kind of like the lenses in which these waking dreams can be experienced. I feel that in my heart of hearts that with Flare HD / ~4K support that this missing ~'Dream Lense' can find its way home after searching for so long and by so many people throughout the decades lost in the dark of the void. It is a hard set of feelings to describe, almost like it is something that should be done to fulfill a larger goal; something greater than one's self.

**Reply by WithinAmnesia** [reply]
Anyway big long form thoughts aside / back down to planet earth XD.

The Flare HDcore is a great step forward. It can even be crudely upscaled in an image editor. Or it could also even be better to use for an UI upscaler for it has more detail to upscale. Admittedly if the Flare art files can be rendered at an enlarged x 4 scale that would be the best solution. I just suck using blender at the moment.

I wonder if the blender script that is used to automatically spit out image frames can be enlarged x 4 in resolution scale compared to the base Flare Fantasy core with its original x 1 scale? I used the default x 1 scale of the art source Blender files in the Fantasy core Github for getting the dual wielding animation frames before. I remember asking Clint over an email a month ago how to adjust the blender animation rig to output enlarged x 4 scale image frames; yet I guess it got eaten in spam or something like before. I did not want to bug Clint too much so I figured maybe someone has tinkered with that before and also know to tweak the blender animation render rig to output enlarged x 4 image frames. If I know what to do I can do the work it is just I am kind of lost in terms what to do with Blender to get the frames rendered bigger. Maybe I have to ask a Blender community or find some other new idea. Or perhaps find more people to hopefully take a look at the blender rig and hopefully find a solution to render the animation frames at the enlarged x 4 scale. If I can get the blender animation render rig to output enlarged x 4 scale image animation frames I can stitch them together with a program as I have done before. I could effectively solo the work load if I knew what to do XD.

**Reply by WithinAmnesia** [reply]
@dorkster Do you know where Clint is describing where the x 2 scale blender files are? 
"The folder art_src contains the raw Blender or Gimp files, or unpacked sprite sheets, unique to the hdcore mod. This folder is not needed during gameplay.

Where the Blender files are not substantially different from fantasycore (e.g. only the render resolution % was changed), those files are not copied here. **_See the flare-game repo, art_src folder for those files._**" - https://github.com/clintbellanger/flare-hdcore

I might be able to reverse engineer the scale difference from x 2 verses x 1 by comparing the two. I hope that there is a x 2 scale blender animation script render rig that outputs the x 2 scale Flare art. If I can figure out how that works I might be able to problem solve a solution for Flare x 4 art assets. I would like help from people who know this but even then if I can find some x 2 blender files and hopefully a x 2 scale render rig script I might be able to pull this off even with minimal blender experience XD. Or maybe just fail upwards which also is improvement XD.

Edit: https://github.com/flareteam/flare-game/tree/master/art_src In here somewhere?

**Reply by WithinAmnesia** [reply]
![image](https://user-images.githubusercontent.com/86871191/151488907-c97dea7d-3a0c-48de-8c52-307a32b433bc.png)
For myself the Flare Game master zip does not include the art_src and tiled folders from the Github Flare game master?

**Reply by dorkster** [reply]
We [exclude](https://github.com/flareteam/flare-game/blob/master/.gitattributes#L11-L13) the art_src and tiled directories in the zip to keep the filesize down for people looking to just get the latest game data. If you want those directories, you'll need to `git clone` the repository. I'm guessing you're not too comfortable with command line stuff, so maybe you can try [Git for WIndows](https://gitforwindows.org/) or Github's own [desktop client](https://desktop.github.com/). I haven't used much of either, so I can't speak much on how to use them.

**Reply by WithinAmnesia** [reply]
![image](https://user-images.githubusercontent.com/86871191/151490091-2839701a-e342-4ec3-8b04-e90587a23749.png)
I figured it out (I think?)! I looked closely at https://clintbellanger.net/articles/isometric_tiles/ to try and figure out what was going on from Clint's perspective. The Blender Render Rig was at x 4 art scale natively / by default but the render scaled it down to 25% (1/4) to make regular Fantasy core animation at assets. So I changed the render scale to 100% and it rendered 'Flare HD' art assets! I was aiming at enlarged x 4 the scale and it turns out that with (weapon animations at least) this is possible with some tweaking XD!

**Reply by WithinAmnesia** [reply]
@dorkster Wow I can use blender (sort of) after all XD. I will see if the also works for tilesets.

**Reply by WithinAmnesia** [reply]
![image](https://user-images.githubusercontent.com/86871191/151491188-c09c2b69-bf43-4e77-b31a-6c71c8cfafa5.png)
It works! That is a Flare HD / Flare 4K resoltion tile! XD!

**Reply by dorkster** [reply]
FYI, stitching together the individual images into a single spritesheet can be done with ImageMagick's `montage` tool.

Here's a example that Clint left in the art_src folder: https://github.com/flareteam/flare-game/blob/master/art_src/characters/hero/montage_script.txt

Obviously, you would change 128x128 to 512x512.

Also, this is what I was talking about in regards to workflow. The Python script that spits out the individual images should also be able to (optionally): combine them, pack them, generate the animation/tileset definition, and move them to the appropriate location in the mods folder.

**Reply by WithinAmnesia** [reply]
You guys have a script for stitching them together? I was using imagemagic montage script shown by Danimal lol XD https://opengameart.org/forumtopic/dual-wielding-animations-and-mechanical-support
"Danimal how do you make your Image Magick montage where they fit a 4096 x 1024 .png where as mine just turn out 2048 x 2016? Which makes a miss match where 2048 times 2 = 4096 (match) but 2016 divided by 2 = 1008 (miss match)? What command line works for you with Image Magic Montage?

I have this so far: magick montage *.png -background none FILENAMEHERE.png

My blender output frame images are 128 x 128 or they should be. I wonder if I have undersized images? Nope I checked and they are all 128 x 128. Then am I using a buggy Image Magic command line? What Image Magick montage command line works then? What one am I supposed to use?

Hmm "PS C:\Desarrollo flare\Terminados\Flying terror> magick montage -background "transparent" -geometry 128x128 -tile 32x8 *.png sprites.png"

Edit: Bingo! 4096 x 1024 sprite tileset for the off hand dagger heroine animations. All that is left to do is test it but it is 7 AM now and I need to go to bed XD. Thank you tremendously

Hopefully I can test this animation successfully later :-P."
+ these files for an Imagemagic Montage for Flare art asset creation manual I made:
[Image Magic Command Line For Animations.txt](https://github.com/flareteam/flare-game/files/7955963/Image.Magic.Command.Line.For.Animations.txt)
![Animation Working Tileset](https://user-images.githubusercontent.com/86871191/151492793-fdf76442-75f5-41a7-8603-8e87e420bfe0.png)

*Wonders, 'How do you guys make montages?' -> looks at link*
https://github.com/flareteam/flare-game/blob/master/art_src/characters/hero/montage_script.txt
Well sh!t lol XD That could have saved some time and like 3 east Indian man youtube video tutorials of how to use Imagemagic montage [they were very informative by the way] and reverse engineering with what Danimal did lol XD. Either way it is looking good so far. Bit by bit XD.

**Reply by WithinAmnesia** [reply]
@dorkster Is there a way to get a script to make large form (non-compact format) working animation and tileset defs that are enlarged x 4 the scale not by hand?

I went nearly bonkers making a tileset def expansion by hand for this (64 x 32 pixel per tile) x 1 scale tileset: https://opengameart.org/content/flare-super-dungeon-tileset-version-10 . Myself and Danimal use the big caveman version of the non-compact tilesets (they are easier to modify and still work and we suck at making things compact XD). Is there a way to use a script to produce character animation tileset defs for these big enlarged x 4 tiles automatically / without going bonkers by hand? Also could this tileset def format be done in the same 16 x 190 format. As like how I made my very large 16 x 190 tile tileset definition (so I can easily keep adding more art tiles in blank area the future)? Is there a way to make a 16 x 190 tileset def for the big form tileset image for the 256 (X axis) x 128 (Y axis) pixel scale by using a script? If so that would be amazing XD. Making a tileset and animation definition by hand is so painful X.x.

I think that the enlarged x 4 large non compact animation total pixel dimensions are:
4096 x 1024 yet multiplied by 4 to create **16384 x 4096** for the character animations.

I think that the enlarged x 4 large non compact 16 x 190 tileset (with enlarged x 4 with 256 x 128 pixel scale tiles) pixel dimensions are: 1024 x 6080 yet multiplied by 4 to create **4096 x 24320** for the 16 x 190 tileset definitions.
So Flare 4K could use scripts to create definitions for:

- **16384 x 4096** pixel dimensions scale for the large non compact format character animations.

- **4096 x 24320** pixel dimensions scale for the large non compact format 16 x 190 tileset definitions.

**Reply by dorkster** [reply]
Sorry, such a script doesn't exist at the moment.

**Reply by WithinAmnesia** [reply]
I think the animation script already works? It is just the tileset def is a bit crazy but it *can* be done by hand for the 4K / 256 x 128 tiles. Although once it is done I can use it as a standard template for the rest of the areas. If I do get the Flare 4K 16 x 190 tileset definition text file made I should share it on Open Game Art to keep up the good will and be a good role model / lead by example / empower the creative community.

**Comment by WithinAmnesia** [comment, marked as answer]
I figured out how to make enlarged x 4 scale Flare 4K high resolution images from changing the render solution from 25% to 100% render resolution in Flare / Fantasy core art blender file source files (that works with the preexisting render script / render rig): https://opengameart.org/forumtopic/what-elements-of-rpgs-do-you-like-post-a-wish-list-of-rpg-features?page=1
"Hold on! I just did a bunch of successful experiments! Check this out: https://github.com/flareteam/flare-game/discussions/895#discussioncomment-2063938 ! I made successful Flare 4K animations for dual wielding arming sword and I also made Flare 4K grassland tilesets with new renders! XD. I looked very close at the screen shots that Clint Bellanger has with the link from [at]bzt here https://clintbellanger.net/articles/isometric_tiles/ and used my Blender 2.79 and the art source files and it turns out the original files were in specification to render at 4 times the size of the regular Flare assets. in fact the regular Flare assets were shrunk down with 25% / 0.25 render scale and by making that render scale go to 100% / 1.00 it produces Flare 4K freaking graphics wow! I might suck with Blender but my detective work was on point and I lucked out lol XD. AAAH I am so relieved to find this HUGE surge of creative power and having the realistic power to do so XD. So far it looks like Flare 4K is coming! https://www.youtube.com/watch?v=6JElrEbAcwY&ab_channel=Sam%26Dave-Topic "

**Comment by WithinAmnesia** [comment]
@dorkster https://github.com/flareteam/flare-game/blob/master/art_src/characters/hero/default_chest.blend I tried to used Blender 2.79 and render using the 8 direction render animation script thing in 100% 512x512 resolution (up from 25% 128x128). Yet the render animations stops at one direction and has a fraction of the frames rendered as it should? What should one do to render the hero and heroine animations at 100% for 512x512 (~Flare X4 / Flare 4K)? I can put the work in for all animations and share the animation art but I have to figure out how to get the default hero and heroines to render at 100% for 512x512 in 8 directions; any suggestions?

**Reply by dorkster** [reply]
I'm sorry, but I can't reproduce the issue you're having. For me, it renders the entire set of animations in all 8 directions (should result in 256 images).

Maybe try looking at the console output? If you're on Windows, you can open the system console by going to `Window > Toggle System Console` within Blender.

**Reply by WithinAmnesia** [reply]
I got it to work. I highlighted the script and right click opened a selection option for 'Run Script' and ran it and it worked with 100% resolution for 256 complete 512 x 512 animation frames. I was previously going render menu -> render animation.

**Reply by WithinAmnesia** [reply]
https://cdn.discordapp.com/attachments/870416511251124254/1052023156145143808/image.png
4K Plate Armour!
